pxtlib/util.ts

```diff
@@ -333,6 +333,115 @@ namespace ts.pxtc.Util {
+                if (!(key in objB)) {
+                    // Gather remove ops
+                    diff.remove.push({
+                        op: 'remove',
+                        path: basePath.concat(key)
+                    });
+                }
+            }
+
+            for (const baseKey of Object.keys(objB)) {
+                const key = Array.isArray(objB) ? Number(baseKey) : baseKey;
+                const atObjA = objA[key];
+                const atObjB = objB[key];
+                if (!(key in objA)) {
+                    // Gather add ops
+                    diff.add.push({
+                        op: 'add',
+                        path: basePath.concat(key),
+                        value: atObjB
+                    });
+                } else if (!Object.is(atObjA, atObjB)) {
+                    // Value exists in both, and they're not the same instance.
+                    if (typeof atObjA !== typeof atObjB) {
+                        // Type changed. Replace the whole subtree.
+                        diff.replace.push({
+                            op: 'replace',
+                            path: basePath.concat(key),
+                            value: atObjB
+                        });
+                    } else {
+                        if (typeof atObjA !== 'object' &&
+                            typeof atObjB !== 'object' &&
+                            atObjA != atObjB) {
+                            // Leaf nodes with differing values. Replace.
+                            diff.replace.push({
+                                op: 'replace',
+                                path: basePath.concat(key),
+                                value: atObjB
+                            });
+                        } else {
+                            // Recurse
+                            diffRecursive(atObjA, atObjB, basePath.concat(key));
+                        }
+                    }
+                }
+            }
+        }
+
+        diffRecursive(outerObjA, outerObjB, []);
+
+        return [...diff.remove.reverse(), ...diff.replace, ...diff.add];
+    }
+
+    /**
+     * Applies a set of JSON Patch operations to the object.
+     */
+    export function jsonPatchInPlace(obj: any, ops: JsonPatchOp[]): void {
+        if (!obj || typeof obj !== 'object') {
+            throw new Error("jsonPatch: Must be an object or an array.");
+        }
+        for (const op of ops) {
+            const path = op.path.slice();
+            const lastKey: any = path.pop();
+            if (lastKey == null) {
+                throw new Error("jsonPath: missing last key");
+            }
+            let subObj = obj;
+            // Add sub-objects for each path element except the last one.
+            let currKey: any;
+            while (((currKey = path.shift())) != null) {
+                if (!(currKey in subObj)) {
+                    subObj[currKey] = {};
```

Comment:
will this work with arrays? should recreate all of the properties, but will they be proper arrays? i.e. can I do subObj[currKey].map()?

Reply:
Zooming out, this feels like something we shouldn't invent. There are edge cases (like creating arrays vs objects, as you point out), and we need unit tests. I generally try to avoid taking additional dependencies, especially client side, but in this case I'd guess that a well-tested version of this algorithm already exists. I'll poke around npm for a good package.

Reply:
Ended up reinventing the wheel anyway. Works with arrays now.